streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala

```diff
@@ -195,30 +236,44 @@ private[streaming] class ReceivedBlockTracker(
       timeToAllocatedBlocks --= batchTimes
     }
 
+    def resolveEvent(event: ReceivedBlockTrackerLogEvent): Unit = {
+      event match {
+        case CombinedReceivedBlockTrackerLogEvent(events) =>
+          events.foreach(resolveEvent)
+        case BlockAdditionEvent(receivedBlockInfo) =>
+          insertAddedBlock(receivedBlockInfo)
+        case BatchAllocationEvent(time, allocatedBlocks) =>
+          insertAllocatedBatch(time, allocatedBlocks)
+        case BatchCleanupEvent(batchTimes) =>
+          cleanupBatches(batchTimes)
+      }
+    }
+
     writeAheadLogOption.foreach { writeAheadLog =>
       logInfo(s"Recovering from write ahead logs in ${checkpointDirOption.get}")
       writeAheadLog.readAll().asScala.foreach { byteBuffer =>
         logTrace("Recovering record " + byteBuffer)
-        Utils.deserialize[ReceivedBlockTrackerLogEvent](
-          byteBuffer.array, Thread.currentThread().getContextClassLoader) match {
-          case BlockAdditionEvent(receivedBlockInfo) =>
-            insertAddedBlock(receivedBlockInfo)
-          case BatchAllocationEvent(time, allocatedBlocks) =>
-            insertAllocatedBatch(time, allocatedBlocks)
-          case BatchCleanupEvent(batchTimes) =>
-            cleanupBatches(batchTimes)
-        }
+        resolveEvent(Utils.deserialize[ReceivedBlockTrackerLogEvent](
+          byteBuffer.array, Thread.currentThread().getContextClassLoader))
       }
     }
   }
 
   /** Write an update to the tracker to the write ahead log */
-  private def writeToLog(record: ReceivedBlockTrackerLogEvent) {
-    if (isWriteAheadLogEnabled) {
-      logDebug(s"Writing to log $record")
+  private[streaming] def writeToLog(record: ReceivedBlockTrackerLogEvent): Boolean = {
+    if (!isWriteAheadLogEnabled) return true
+    if (WriteAheadLogUtils.isBatchingEnabled(conf)) {
+      val promise = Promise[Boolean]()
+      walWriteStatusMap.put(record, promise)
+      walWriteQueue.offer(record)
+      Await.result(promise.future.recover { case _ => false }(batchWriterThreadPool),
```

Comment:
we need to add a try-catch block with `TimeoutException` here which returns `false`